### ISSUE TITLE
nil.singleton_method does not find method defined on nil

### ISSUE THREAD

**Issue by headius** [jruby/jruby#7787]
Investigating this with @jeremyevans I discovered JRuby has some broken behavior around methods defined on nil:

```
[] jruby $ jruby -e 'def nil.foo; end; p nil.singleton_method(:foo)'
NameError: undefined method `foo' for `nil'
  singleton_method at org/jruby/RubyKernel.java:2319
            <main> at -e:1
[] jruby $ (chruby 3.2 ; ruby -e 'def nil.foo; end; p nil.singleton_method(:foo)')
#<Method: NilClass#foo() -e:1>
```

This behavior does not match CRuby, but CRuby also is not consistent here in that `singleton_methods` does **not** see the method:

https://bugs.ruby-lang.org/issues/11064

I believe it should consistently see the method or consistently not see the method, but a decision would have to be made on the CRuby side. In the short term, we have never had this filed as an issue, so I'm inclined to leave it alone until such time as ruby-core decides it should be visible through these methods.

**Comment by sampokuokkanen** [comment]
Seems like CRuby actually now agrees with JRuby and we can close this issue! The tests added to ruby/spec also pass on JRuby as is and there are no exclusions for them. 
https://github.com/ruby/ruby/pull/7973

**Comment by headius** [comment]
OMG I didn't think to retest CRuby. Indeed, it errors.

I'd love to get @jeremyevans input here but I agree, we can close this.

**Comment by jeremyevans** [comment]
Yep we can close this. This is one of the bugs we investigated during the RubyKaigi 2023 Rubyists on Rails train between Tokyo and Matsumoto. Bug was originally reported back in 2015 (https://bugs.ruby-lang.org/issues/11064). :)
